pipelines/pipedrive/__init__.py

```diff
@@ -10,151 +10,108 @@
 """
 
 import dlt
-import functools
+from typing import Any, Iterator, Optional, Sequence
+from functools import partial
 
-from .custom_fields_munger import munge_push_func, pull_munge_func, parsed_mapping
+from .custom_fields_munger import parsed_mapping
+from .helpers import get_recents, get_endpoint
 from dlt.common.typing import TDataItems
+from dlt.extract.typing import TWriteDisposition
 from dlt.extract.source import DltResource
-from dlt.sources.helpers import requests
-from typing import Any, Dict, Iterator, Optional, Sequence
+from dlt.extract.typing import TableNameMeta
 
 
 @dlt.source(name='pipedrive')
-def pipedrive_source(pipedrive_api_key: str = dlt.secrets.value) -> Sequence[DltResource]:
-    """
+def pipedrive_fields_source(
+    pipedrive_api_key: str = dlt.secrets.value
+) -> Sequence[DltResource]:
+    # we need to add entity fields' resources before entities' resources in order to let custom fields data munging work properly
+    entity_fields_endpoints = ['activityFields', 'dealFields', 'organizationFields', 'personFields', 'productFields']
+    resources =  [dlt.resource(get_endpoint(endpoint, pipedrive_api_key), name=endpoint, write_disposition='replace') for endpoint in entity_fields_endpoints]
+    resources.append(parsed_mapping())
+    return resources
 
+
+@dlt.source(name='pipedrive')
+def pipedrive_source(
+    pipedrive_api_key: str = dlt.secrets.value, entities: Sequence[str] = (), write_disposition: TWriteDisposition = 'merge'
+) -> Sequence[DltResource]:
+    """
     Args:
-    pipedrive_api_key: https://pipedrive.readme.io/docs/how-to-find-the-api-token
+        pipedrive_api_key: https://pipedrive.readme.io/docs/how-to-find-the-api-token
+        entities: Optional list of entity names to load. By default all entities returned by `/recents` endpoint are included
+            See https://developers.pipedrive.com/docs/api/v1/Recents#getRecents for possible entities
 
     Returns resources:
-        activityFields
-        dealFields
-        deals
+        recent_items
```

Comment:
All these entities are included in /recents so it's all just one resource now creating multiple tables like in GH pipeline.   
But maybe separate resources are better so you can use with_resources and such. Should be doable with transform per each entity type while still loading recents only once.

Reply:
What we want is
1. To load all the endpoints that we had listed
2. To use "recents" as a way to provide incremental loading for the entities supported by Recents.

Unfortunately the recents endpoint was not clearly documented so I did not understand how to best use it. According to the docs, it "Returns data about all recent changes occurred after the given timestamp."

It seems instead this endpoint does not return "information about all recent changes" but it returns "Latest state of entities that were changed since the requested timestamp". Assuming the requested timestamp is not limited to "recent", then this means we could use this endpoint to just get the latest state whenever (since all time for first load, since last load for subsequent loads)

So it seems we have 2 types of resources with 2 types of "loader methods"
1. Recents - First load can load "all time" data, second load can load the newly changed ones with "merge". To enable selecting them as resource we could request per entity from this endpoint. The original idea here was to get the full load form the entity endpoint and the "recents" merged from recents (in line with what docs suggests recents returns).

2. Stateful entities with "replace" write disposition - everything else. Most of the endpoints can be requested with no extra parameters, some need to be requested with extra params, some need dependent entity.

Reply:
should we have two source them? one for a full load and second one (a merge) that you use to load incrementally starting from a timestamp? (ie. last value - one day?) or you want to have a single source where we have everything?

Reply:
@adrianbr could you take a look? this will unblock @steinitzu

Reply:
> should we have two source then? one for a full load (use case 1) and second one (a merge on recent) that you use to load incrementally starting from a timestamp? (ie. last value - one day?) or you want to have a single source where we have everything?

A single source.